From bea5ca1141531a4c2c14c4059ebfcf0b00862d42 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Thu, 26 May 2016 19:53:21 +0000
Subject: [PATCH] [CONFIGURATION-632] Better support for interpolation with
 arrays.

Variables that reference properties with multiple values are now
handled correctly. The getList() and getStringArray() methods then
return all values of the referenced property.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1745638 13f79535-47bb-0310-9956-ffa450edef68
---
 .../configuration2/CompositeConfiguration.java  |  4 ++--
 .../TestCompositeConfiguration.java             | 17 +++++++++++++++++
 2 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/src/main/java/org/apache/commons/configuration2/CompositeConfiguration.java b/src/main/java/org/apache/commons/configuration2/CompositeConfiguration.java
index f23eb9ba7d..91b8da228f 100644
--- a/src/main/java/org/apache/commons/configuration2/CompositeConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/CompositeConfiguration.java
@@ -597,10 +597,10 @@ private void replaceInMemoryConfiguration(Configuration config)
      * @param config the configuration to query
      * @param key the key of the property
      */
-    private static void appendListProperty(List<Object> dest, Configuration config,
+    private  void appendListProperty(List<Object> dest, Configuration config,
             String key)
     {
-        Object value = config.getProperty(key);
+        Object value = interpolate(config.getProperty(key));
         if (value != null)
         {
             if (value instanceof Collection)
diff --git a/src/test/java/org/apache/commons/configuration2/TestCompositeConfiguration.java b/src/test/java/org/apache/commons/configuration2/TestCompositeConfiguration.java
index 236f50b601..66608454ce 100644
--- a/src/test/java/org/apache/commons/configuration2/TestCompositeConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/TestCompositeConfiguration.java
@@ -17,6 +17,7 @@
 
 package org.apache.commons.configuration2;
 
+import static org.junit.Assert.assertArrayEquals;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
@@ -794,6 +795,22 @@ public void testInterpolationInMultipleConfigs()
                 cc.getString("property.one.ref"));
     }
 
+    /**
+     * Tests whether interpolation works if a variable references a property
+     * with multiple values. This test is related to CONFIGURATION-632.
+     */
+    @Test
+    public void testInterpolationArrayReference()
+    {
+        Configuration props = new PropertiesConfiguration();
+        String[] values = { "a", "property", "with", "multiple", "values" };
+        props.addProperty("keyMultiValues", values);
+        props.addProperty("keyReference", "${keyMultiValues}");
+        cc.addConfiguration(props);
+        assertArrayEquals("Wrong interpolated value", values,
+                cc.getStringArray("keyReference"));
+    }
+
     /**
      * Tests the behavior of setListDelimiterHandler() if the in-memory configuration
      * is not derived from BaseConfiguration. This test is related to
